### PR TITLE
src/testdir: use -u unix.vim, align with Vim's behavior

### PR DESCRIPTION
**Pull request by blueyed**
This also removes the unnecessary `-U NONE` for us, given that there is
a diff there anyway already (not using
`NO_PLUGIN` (added in 7.4.1420), `NO_PLUGINS`/`NO_INITS` (new in this patch)).

Ref: https://github.com/neovim/neovim/issues/10649#issuecomment-515864782

TODO:

- [x] https://github.com/neovim/neovim/pull/10197
- [x] https://github.com/neovim/neovim/pull/10651
- [ ] keep https://github.com/neovim/neovim/commit/95979afc47f33ddde158cf699c1112a7be5cfc47 (partly)?

### CODE REVIEW

**Comment by blueyed** [comment]
This fails because of `directory=.` added via https://github.com/neovim/neovim/commit/95979afc47f33ddde158cf699c1112a7be5cfc47.

File: src/nvim/testdir/unix.vim [hunk 1]
```diff
@@ -10,6 +10,3 @@ endif
 
 " Don't depend on system locale, always use utf-8
 set encoding=utf-8
-
-" Use safer defaults for various directories
-set backupdir=. directory=. undodir=. viewdir=.
```
Comment (justinmk) [inline src/nvim/testdir/unix.vim]: User files could be overwritten

Comment (blueyed) [reply src/nvim/testdir/unix.vim]: Yes.. it was more meant as a test to see how this behaves when removed etc.

In general I think `set noswapfile` makes more sense than `directory=.`.
Otherwise it should follow a safe setting of `$HOME`, shouldn't/doesn't it?

**Comment by blueyed** [comment]
So.. I suggest we get rid of `unix.vim`, and instead only have a custom section in `setup.vim`, which can also check `win32` still etc.
`unix.vim` was already used as a means of "do this for setup", and it is rather confusing.

After all `-u NONE` makes sense / is required to have nvim not do syntax highlighting by default for example.

**Comment by blueyed** [comment]
Vim basically uses `unix.vim` (or `dos.vim` etc) as entrypoints for `setup.vim`, which we can skip.

File: src/nvim/testdir/unix.vim (hunk 1 above)
Comment (blueyed) [reply src/nvim/testdir/unix.vim]: What do you suggest since we cannot have `set directory=.` as is?
Leave only that out for now?
(Related: https://github.com/vim/vim/pull/4742)

Comment (justinmk) [reply src/nvim/testdir/unix.vim]: can it be set a a temp dir?

**Comment by blueyed** [comment]
oldtest failures on AppVeyor look the same.

File: src/nvim/testdir/unix.vim [hunk 2]
```diff
@@ -2,14 +2,20 @@
...
+source setup.vim
+
+" TODO: upstream?
+" Initially added in https://github.com/neovim/neovim/commit/4a5bc6275d090.
+" Appears to make sense as in "adjust for set shell=sh".
 if has('win32')
   set shellcmdflag=-c shellxquote= shellxescape= shellquote=
   let &shellredir = '>%s 2>&1'
-  set shellslash
 endif
```
Comment (blueyed) [inline src/nvim/testdir/unix.vim]: Happy to suggest this upstream, if you think it / my comment with it makes sense.  /cc @janlazo

**Review by justinmk — APPROVED** [review]

File: src/nvim/testdir/unix.vim (hunk 2 above)
Comment (janlazo) [reply src/nvim/testdir/unix.vim]: Sure. It makes sense to me.

File: src/nvim/testdir/unix.vim [hunk 3]
```diff
@@ -2,14 +2,20 @@
...
+
+source setup.vim
+
+" TODO: upstream?
+" Initially added in https://github.com/neovim/neovim/commit/4a5bc6275d090.
+" Appears to make sense as in "adjust for set shell=sh".
 if has('win32')
   set shellcmdflag=-c shellxquote= shellxescape= shellquote=
   let &shellredir = '>%s 2>&1'
-  set shellslash
```
Comment (janlazo) [inline src/nvim/testdir/unix.vim]: Why is this removed? This shouldn't break anything and `shellslash` goes together with `sh` shell.

File: src/nvim/testdir/setup.vim [hunk 4]
```diff
@@ -1,10 +1,61 @@
...
+  endif
+endif
+
+" Nvim: we are using "-u NONE" (to disable syntax etc reliably).
+" But keeping unix.vim is good in general (mentioned in help), and required
+" for oldesttest likely still.
+if exists('+shellslash')
+  " Native Windows build.
+  " Use default shell on Windows to avoid segfault, caused by TUI
+  " TODO: revisit?!
```
Comment (janlazo) [inline src/nvim/testdir/setup.vim]: TUI shouldn't segfault anymore but the code that follows should be kept because of `SHELL`.

File: src/nvim/testdir/setup.vim [hunk 5]
```diff
@@ -1,10 +1,61 @@
...
+  let $SHELL = ''
+  let $TERM = ''
+  let &shell = empty($COMSPEC) ? exepath('cmd.exe') : $COMSPEC
+  set shellcmdflag=/s/c shellxquote=\" shellredir=>%s\ 2>&1
+  let &shellpipe = &shellredir
+
+  " TODO/XXX: Vim has this in dos.vim (for MingW also):
+  " " Settings for test script execution
+  " " Always use "COMMAND.COM", don't use the value of "$SHELL".
+  " set shell=c:\COMMAND.COM shellquote= shellxquote= shellcmdflag=/c shellredir=>
```
Comment (janlazo) [inline src/nvim/testdir/setup.vim]: Vim should detect `COMSPEC` instead which then overwrites the value of `shell` on Windows. This hack shouldn't be necessary, at least for Neovim.

I haven't seen anyone, besides myself, use `COMMAND.COM` over `cmd.exe` on Windows. It exists only for backward compatibility at this point.

File: src/nvim/testdir/runtest.vim [hunk 6]
```diff
@@ -64,7 +64,7 @@ set nomore
 lang mess C
 
 " Always use forward slashes.
-" set shellslash
+set shellslash
```
Comment (janlazo) [inline src/nvim/testdir/runtest.vim]: I'd hold back on this because oldtests failures changed and it affects `shellescape` for cmd.exe but oldtests don't fail the Appevyor builds.

Comment (erw7) [reply src/nvim/testdir/runtest.vim]: It is possible to change an existing test if the specification changes or if there is a problem with the test itself.

The changes that commented out `set shellslash` in https://github.com/neovim/neovim/commit/79a8d905ab8908362109284240d2f9a75564f8ad seem correct at first glance, but that is a mistaken change, because vim uses `shellslash` if `cmd.exe` is the shell(In fact, we commented this out, so in case of `set shellslash`, the problem that the file expansion is wrong is missed). Thus, the change to re-enable `set shellslash` is the correct change.

If you want test to succeed, you should change it to run with `set noshellslash` with the following comment for each individual test that fails:

```vim
" FIXME: Because the test fails by issue #xxxx, set noshellslash and test execution
" FIXME: Run with set noshellslash because test fails with set shellslash
```

File: src/nvim/testdir/unix.vim (hunk 3 above)
Comment (blueyed) [reply src/nvim/testdir/unix.vim]: Will remove it completely - according to https://github.com/vim/vim/pull/4749 it should not be necessary (i.e. it should follow "set shell=sh").

Comment (blueyed) [reply src/nvim/testdir/unix.vim]: ..but nvim does not have that: https://github.com/vim/vim/blob/8c5a278fc508da6dfe50e69b6ee734451aa4eafb/src/option.c#L4173-L4232

Comment (blueyed) [reply src/nvim/testdir/unix.vim]: Will move it to setup.vim

**Comment by blueyed** [comment]
So I think it makes sense to get rid of `unix.vim` for us.

As for `set shell=sh` with Windows: we should probably have a way to configure this, e.g. `$NVIM_TEST_SHELL`, and then also use this on CI.  AFAICS we're not using both variants (`shell=sh`, default shell (i.e. `cmd.exe`) on AppVeyor, are we?

**Comment by justinmk** [comment]
> So I think it makes sense to get rid of `unix.vim` for us.

Yes.

> As for `set shell=sh` with Windows: we should probably have a way to configure this, e.g. `$NVIM_TEST_SHELL`, and then also use this on CI. 

Can we avoid more env vars? Why can't it switch on `has('win32')` in setup.vim (or whatever)? 

Would like to just move forward with this if it improves things without breaking other things.

**Comment by blueyed** [comment]
> Why can't it switch on `has('win32')` in setup.vim (or whatever)?

AFAIK `has('win32')` is true for all jobs on AppVeyor?
My point here is that we should probably default to `cmd.exe` there, but might also want to simulate the other case (Windows users with Unix settings).

**Comment by janlazo** [comment]
neovim detects `$SHELL` to set `shell`. Is that enough to detect which shell to use?

**Comment by erw7** [comment]
@blueyed `unix.vim` is also used in the following places. I think you need to change it to `setup.vim`. `unix.vim` of neovim does not have `source setup.vim`, but [`unix.vim`](https://github.com/vim/vim/blob/088e8e3443520dec91a384081e66445a104810bb/src/testdir/unix.vim#L13) and [`dos.vim`](https://github.com/vim/vim/blob/088e8e3443520dec91a384081e66445a104810bb/src/testdir/dos.vim#L9) of vim has `source setup.vim`.

https://github.com/neovim/neovim/blob/25fff17d19f5927987526bd4d7c4bd4b73603241/src/nvim/testdir/Makefile#L100

**Comment by justinmk** [comment]
> My point here is that we should probably default to `cmd.exe` there, but might also want to simulate the other case (Windows users with Unix settings).

I don't know, is this really important right now? I'd rather not spend time thinking about / enhancing the testdir/ tests if it means diverging from upstream unnecessarily.

File: src/nvim/testdir/setup.vim [hunk 7]
```diff
@@ -23,25 +70,6 @@ set wildoptions=
...
-if has('win32')
-  let $SHELL = ''
-  let $TERM = ''
-  let &shell = empty($COMSPEC) ? exepath('cmd.exe') : $COMSPEC
-  set shellcmdflag=/s/c shellxquote=\" shellredir=>%s\ 2>&1
-  let &shellpipe = &shellredir
-endif
+" Don't depend on system locale, always use utf-8.
+" Ref: https://github.com/neovim/neovim/pull/2929
+set encoding=utf-8
```
Comment (justinmk) [inline src/nvim/testdir/setup.vim]: `set encoding` does nothing in Nvim. The reference to https://github.com/neovim/neovim/pull/2929 is outdated, that PR was before we made 'encoding' readonly.

So this line isn't needed unless it's in Vim.

**Comment by justinmk** [comment]
> This fails because of `directory=.` added via [95979af](https://github.com/neovim/neovim/commit/95979afc47f33ddde158cf699c1112a7be5cfc47).

How is it passing now? 

@blueyed do you want to squash this somehow? Anything remaining to be done?

**Comment by blueyed** [comment]
> > This fails because of `directory=.` added via [95979af](https://github.com/neovim/neovim/commit/95979afc47f33ddde158cf699c1112a7be5cfc47).
> 
> How is it passing now?

This is still removed here:
```
" Use safer defaults for various directories
set backupdir=. directory=. undodir=. viewdir=.
```

Should use a temp dir maybe instead like you suggested (https://github.com/neovim/neovim/pull/10650#discussion_r308168635).

> @blueyed do you want to squash this somehow?

Sure.

> Anything remaining to be done?

Will have to go through it again - e.g. with regard to the above.
But it is looking quite good altogether I think.

File: Makefile [hunk 8]
```diff
@@ -122,6 +122,10 @@ else
 	@# NOTE: supports TEST_FILE=test_foo.res to run test_foo actually
 	+$(SINGLE_MAKE) -C src/nvim/testdir NVIM_PRG="$(realpath build/bin/nvim)" $(patsubst %.res,%,$(TEST_FILE)) SCRIPTS= $(MAKEOVERRIDES)
 endif
+# Build oldtest by specifying the .vim filename.
+.PHONY: phony_force
+src/nvim/testdir/%: phony_force
+	$(MAKE) oldtest TEST_FILE=$(patsubst src/nvim/testdir/%.vim,%,$@)
```
Comment (justinmk) [inline Makefile]: this is nice, but it would be better to have `TEST_FILE=foo.vim` working instead, so that we can have simpler documentation and consistent usage.

Comment (blueyed) [reply Makefile]: The rule here handles "src/…/test_foo.vim" => TEST_FILE=test_foo (removing the .vim).
The ".res" above is substituted for B/C.
`make src/nvim/testdir/test_environ.res` does not work.  It could be made to work, but it is better to not have it.
I will change the pattern above to only handle `%.vim`.

Comment (justinmk) [reply Makefile]: > `make src/nvim/testdir/test_environ.res` does not work. It could be made to work, but it is better to not have it.

I'm not asking for `make src/nvim/testdir/test_environ.res` to work. I meant `TEST_FILE=foo.vim` specifically. Currently (on master) this works:

    TEST_FILE=foo.res make oldtest

But this would be better (more intuitive):

        TEST_FILE=foo.vim make oldtest

Comment (blueyed) [reply Makefile]: Ok, should work now also.  It is a bit unfortunate that "TEST_FILE" is used here.

After all `make -C src/nvim/testdir test_foo` now also works - but I've kept it to go through the old rule for dependencies etc.  But it also makes it very noisy.

Comment (blueyed) [reply Makefile]: > but I've kept it to go through the old rule for dependencies etc. But it also makes it very noisy.

...which appears to be not working properly anymore for the TSan build.

**Comment by blueyed** [comment]
The tsan job appears to be affected by what https://github.com/neovim/neovim/pull/10479 fixes (likely the help tags are not built there anymore already as a side-effect).  Will rebase it after #10479.

File: test/README.md [hunk 9]
```diff
@@ -79,9 +79,8 @@ To run all legacy Vim tests:
 
 To run a *single* legacy test set `TEST_FILE`, for example:
 
-    TEST_FILE=test_syntax.res make oldtest
+    TEST_FILE=test_syntax make oldtest
```
Comment (justinmk) [inline test/README.md]: if `TEST_FILE=test_syntax.vim` works then it should recommend that. This clarifies for new contributors the separation between  "vimscript tests" and  "lua tests".

**Comment by justinmk** [comment]
AV failure looks unrelated, rebase?

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
Too outdated, likely not longer needed.
